### DISCUSSION TITLE
How to test a ipxe-script with qemu?

### DISCUSSION THREAD

**Discussion by guettli** [ipxe/ipxe#1077] [General]
I want to test an ipxe-script with qemu.

Up to now, I only found solutions which explain how to create a custom iso and then load this.

Is there no way to test an ipxe-script directly by using the default ipxe.iso image?

**Comment by guettli** [comment]
I found this solution:

```
❯ qemu-system-x86_64 -boot n -m 512 -cdrom ipxe.iso -k de  -device virtio-net,netdev=n1 -netdev user,id=n1,tftp=$(pwd),bootfile=/alpine.ipxe 
```

in the same directory I created the file `alpine.ipxe`. This seems to create a tftp server in `pwd`.

**Reply by NiKiZe** [reply]
Yes you can specify bootfile which generates a dhcp response with filename set.
Either this or configure your own dhcp server to send the options as explained at https://ipxe.org/howto/chainloading

**Comment by guettli** [comment]
But now it fails like this:

![image](https://github.com/ipxe/ipxe/assets/414336/125ffbc9-59a4-42d3-a758-0e68b98701f0)

**Reply by NiKiZe** [reply]
You can not chain isos directly, you could try sanboot, however your boot process would probably fail in a later stage.
You will want to follow up on how to PXE boot the OS you want to boot.
for iPXE it is recommended to extract kernel and helper files etc, See https://boot.ipxe.org/demo/ for example.

**Reply by guettli** [reply]
Thank you for your answer.

I could start Alpine via qemu this way:

file: alpine.ipxe
```
#!ipxe

# Unfortunately "https" will fail with current iPXE with "operation not permitted"
# https://ipxe.org/err/410de1
set base-url http://dl-cdn.alpinelinux.org/alpine/v3.18/releases/x86_64/netboot-3.18.4

kernel ${base-url}/vmlinuz-virt console=tty0 modules=loop,squashfs quiet nomodeset alpine_repo=https://dl-cdn.alpinelinux.org/alpine/v3.18/main modloop=${base-url}/modloop-virt
initrd ${base-url}/initramfs-virt
boot
```

```
❯ qemu-system-x86_64 -boot n -m 4000 -k de  -device virtio-net,netdev=n1 -netdev user,id=n1,tftp=$(pwd),bootfile=/alpine.ipxe
```

![image](https://github.com/ipxe/ipxe/assets/414336/b1faea0d-ca83-4ec1-85cc-7b03d17c7bc1)


![image](https://github.com/ipxe/ipxe/assets/414336/edc1c58e-89c2-48a8-b4ae-0566f5d3f019)
